### PR TITLE
Add bindingContext function to FileProcessListener - #2872

### LINKED ISSUE #2872: Support for BindingContext in FileProcessListener
## Current Behavior
Implementations of `Rule` have access to the `BindingContext`, granting them the ability to perform complex analysis within the scope of the entire compilation unit. On the other hand, the `FileProcessListener` (and `AbstractProcessor`, etc) APIs do not provide any way to access the `BindingContext`, limiting their ability to resolve types, functions, and so forth.

## Expected Behavior
In order to enable the same level of analysis for implementations of `FileProcessListener`, the API should provide them a `BindingContext` against which references can be resolved.

## Context
I am in the process of implementing "occurrences counting" metrics for our codebase such as "number of invocations of function A" and "number of constructor calls to classes which implement interface X". Complex queries such as this require a `BindingContext`, but the provided mechanism for making such queries, `AbstractProjectMetricProcessor`, does not provide it.

### PR DESCRIPTION
**Pull request by arturbosch**
Closes #2872

### CODE REVIEW

File: detekt-api/src/main/kotlin/io/gitlab/arturbosch/detekt/api/FileProcessListener.kt [hunk 1]
```diff
@@ -13,25 +18,79 @@ interface FileProcessListener : Extension {
      * Use this to gather some additional information for the real onProcess function.
      * This calculation should be lightweight as this method is called from the main thread.
      */
-    fun onStart(files: List<KtFile>) {}
+    @Deprecated(
+        "Use alternative with a binding context.",
+        ReplaceWith("onStart(files, bindingContext)")
+    )
+    fun onStart(files: List<KtFile>) {
```
Comment (cortinico) [inline detekt-api/src/main/kotlin/io/gitlab/arturbosch/detekt/api/FileProcessListener.kt:25]: nit: is there a specific reason why there is a new line between the parentheses here?

File: detekt-api/src/main/kotlin/io/gitlab/arturbosch/detekt/api/FileProcessListener.kt [hunk 2]
```diff
@@ -1,10 +1,15 @@
+@file:Suppress("DEPRECATION")
```
Comment (BraisGabin) [inline detekt-api/src/main/kotlin/io/gitlab/arturbosch/detekt/api/FileProcessListener.kt:1]: Maybe inline this Suppress? This way, when we remove the old versions we will remove the Suppress too for sure. To ensure that we don't forget to remove this.

**Comment by schalkms** [comment]
We should discuss when we want to have settings for type and symbol solving as a mandatory step. This would reduce the complexity of detekt.

File: detekt-api/src/main/kotlin/io/gitlab/arturbosch/detekt/api/FileProcessListener.kt (hunk 1 above)
Comment (arturbosch) [reply detekt-api/src/main/kotlin/io/gitlab/arturbosch/detekt/api/FileProcessListener.kt:25]: IntelliJ auto formatting ... idk if they updated it or we never formatted it accordingly

**Comment by arturbosch** [comment]
> We should discuss when we want to have settings for type and symbol solving as a mandatory step. This would reduce the complexity of detekt.

This could be quite hard for users with generated code to configure properly. I will soon invest more time in the compiler plugin where we get resolution for free.

**Pull request merged**